### PR TITLE
nn_graph label smoothing

### PR DESCRIPTION
**Pull request by CPFLAME**
### 这个pr要解决的：
- [x] 在nn_graph中加入label smooth的trick进行训练，训练结果见下图
- [x] 和pytorch实现的label smooth进行对齐

pytorch的实现源自于[此处](https://github.com/pytorch/pytorch/issues/7455#issuecomment-513062631)
运行python utils/test_labelsmooth.py即可得到结果，误差最大值的量级在e-7, 几乎一样

```
abs diff: 1.1920928955078125e-07, flow labelsmooth loss: 1.3462207317352295, torch labelsmooth loss:1.34622061252594
abs diff: 1.1920928955078125e-07, flow labelsmooth loss: 1.5114572048187256, torch labelsmooth loss:1.511457085609436
abs diff: 0.0, flow labelsmooth loss: 1.6171989440917969, torch labelsmooth loss:1.6171989440917969
abs diff: 0.0, flow labelsmooth loss: 0.6735591292381287, torch labelsmooth loss:0.6735591292381287
abs diff: 0.0, flow labelsmooth loss: 1.9999173879623413, torch labelsmooth loss:1.9999173879623413
abs diff: 1.1920928955078125e-07, flow labelsmooth loss: 1.3488662242889404, torch labelsmooth loss:1.3488661050796509
abs diff: 0.0, flow labelsmooth loss: 1.5050686597824097, torch labelsmooth loss:1.5050686597824097
abs diff: 1.1920928955078125e-07, flow labelsmooth loss: 1.449904203414917, torch labelsmooth loss:1.4499040842056274
abs diff: 0.0, flow labelsmooth loss: 1.8043279647827148, torch labelsmooth loss:1.8043279647827148
abs diff: 5.960464477539063e-08, flow labelsmooth loss: 0.7393254041671753, torch labelsmooth loss:0.7393253445625305
```

#### 添加loss后结果比较
共用超参
- 数据集：imagenette
- lr: 0.001
- batch_size: 32
- optimizer: sgd
- momentum: 0.9
- epochs: 20

不同超参：

- ce_loss: 采用cross_entropy loss
- ls_loss: 采用label_smooth loss
在nn_graph中使用ce_loss(cross_entropy loss)和ls_loss(label_smooth loss)对比结果：
#### val acc比较
compare_ce_ls_acc
#### loss数值比较
compare_ce_ls_loss

由于数据集比较小，结果可能存在偶然性，但是根据对比得出label smooth可以正常训练，目前看来没有什么问题

### CODE REVIEW

File: resnet50/check/check_labelsmoothing.py [hunk 1]
```diff
@@ -0,0 +1,409 @@
...
+    
+    def forward(self, input, label):
+        onehot_label = flow.F.one_hot(label, num_classes= self.num_classes, 
+                                                on_value=1-self.smooth_rate, 
+                                                off_value=self.smooth_rate/(self.num_classes-1))
+        log_prob = input.softmax(dim=-1).log()
+        onehot_label = onehot_label.to(log_prob.dtype).to("cuda")
+        loss = flow.mul(log_prob * -1, onehot_label).sum(dim=-1).mean()
+        return loss
+
```
Comment (CPFLAME) [inline resnet50/check/check_labelsmoothing.py:60]: 在此处定义label smooth函数

File: resnet50/graph/train_labelsmooth.py [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+        self.smooth_rate = smooth_rate
+    
+    def forward(self, input, label):
+        onehot_label = flow.F.one_hot(label, num_classes= self.num_classes, 
+                                                on_value=1-self.smooth_rate, 
+                                                off_value=self.smooth_rate/(self.num_classes-1))
+        log_prob = input.softmax(dim=-1).log()
+        onehot_label = onehot_label.to(log_prob.dtype).to("cuda")
+        loss = flow.mul(log_prob * -1, onehot_label).sum(dim=-1).mean()
+        return loss
```
Comment (CPFLAME) [inline resnet50/graph/train_labelsmooth.py:57]: 在此处定义label smooth

File: resnet50/graph/train_labelsmooth.py [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+    resnet50_module = resnet50()
+    if args.load_checkpoint != "":
+        print("load_checkpoint >>>>>>>>> ", args.load_checkpoint)
+        resnet50_module.load_state_dict(flow.load(args.load_checkpoint))
+
+    end_t = time.time()
+    print("init time : {}".format(end_t - start_t))
+
+    # of_cross_entropy = flow.nn.CrossEntropyLoss()
+    of_loss = LabelSmoothLoss(num_classes=args.num_classes, smooth_rate=args.label_smooth_rate)
```
Comment (CPFLAME) [inline resnet50/graph/train_labelsmooth.py:85]: 初始化

File: resnet50/graph/train_labelsmooth.py [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+            self.of_loss = of_loss
+            # self.add_optimizer("sgd", of_sgd)
+            self.add_optimizer(of_sgd)
+            self.train_data_loader = train_data_loader
+        
+        def build(self):
+            image, label = self.train_data_loader()
+            image = image.to("cuda")
+            logits = self.resnet50(image)
+            loss = self.of_loss(logits, label)
```
Comment (CPFLAME) [inline resnet50/graph/train_labelsmooth.py:107]: 直接调用即可

File: resnet50/utils/test_labelsmoothloss.py [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+        loss = LabelSmooth(num_classes=3, smooth_rate=0.1)
+        flow_loss = loss(flow_input, flow_label)
+
+        #pytorch
+        torch_input = torch.from_numpy(np_input.copy()).to(torch.float32)
+        torch_label = torch.from_numpy(np_label.copy()).to(torch.int64)
+        loss = LabelSmoothingLoss(classes=3, smoothing=0.1)
+        torch_loss = loss(torch_input, torch_label)
+
+        print(f"abs diff: {flow_loss.numpy() - torch_loss.numpy()}, flow labelsmooth loss: {flow_loss}, torch labelsmooth loss:{torch_loss}")
```
Comment (yuanms2) [inline resnet50/utils/test_labelsmoothloss.py:60]: 最后一行换行符
